src/redisearch_rs/encode_decode/src/qint.rs

```diff
@@ -0,0 +1,284 @@
+    for (i, &value) in values.iter().enumerate() {
+        ret += qint_encode_stepwise(&mut leading, cursor, value, i as u32)?;
+    }
+    cursor.seek(SeekFrom::Start(pos))?;
+    cursor.write_all(&[leading])?;
+    Ok(ret)
+}
+
+/// Decodes a QInt buffer into an array of integers
+///
+/// # Arguments
+/// * `reader` - Buffer reader
+/// * `N` - Number of integers to decode (2, 3, or 4)
+///
+/// # Returns
+/// A tuple of (decoded_values as an array, bytes_consumed) or an io error
+pub(crate) fn qint_decode<const N: usize, R: Read>(
+    reader: &mut R,
+) -> Result<([u32; N], usize), std::io::Error> {
+    // Ensure N is valid (2, 3, or 4)
+    assert!((2..=4).contains(&N), "N must be 2, 3, or 4");
+
+    let mut total = 0;
+
+    // Read the leading byte
+    let mut leading = [0; 1];
+    reader.read_exact(&mut leading)?;
+    total += 1;
+
+    // Decode N values based on 2-bit fields in the leading byte
+    let mut result = [0u32; N];
+    for (i, item) in result.iter_mut().enumerate().take(N) {
+        // Extract 2-bit field for the i-th value
+        let bits = (leading[0] >> (i * 2)) & 0x03;
+        let (val, bytes) = qint_decode_value(bits, reader)?;
+        *item = val;
+        total += bytes;
+    }
+
+    Ok((result, total))
+}
+
+/// Encodes 4 integers into a QInt buffer
+///
+/// # Arguments
+/// * `cursor` - Buffer writer
+/// * `a` - First integer
+/// * `b` - Second integer
+/// * `c` - Third integer
+/// * `d` - Fourth integer
+///
+/// # Returns
+/// The number of bytes written to the buffer or an io error
+pub fn qint_encode4<W>(cursor: &mut W, a: u32, b: u32, c: u32, d: u32) -> Result<usize, io::Error>
+where
+    W: Write + Seek,
+{
+    qint_encode(cursor, [a, b, c, d])
+}
+
+/// Encodes 3 integers into a QInt buffer
+///
+/// # Arguments
+/// * `cursor` - Buffer writer
+/// * `a` - First integer
+/// * `b` - Second integer
+/// * `c` - Third integer
+///
+/// # Returns
+/// The number of bytes written to the buffer or an io error
+pub fn qint_encode3<W>(cursor: &mut W, a: u32, b: u32, c: u32) -> Result<usize, io::Error>
+where
+    W: Write + Seek,
+{
+    qint_encode(cursor, [a, b, c])
+}
+
+/// Encodes 2 integers into a QInt buffer
+///
+/// # Arguments
+/// * `cursor` - Buffer writer
+/// * `a` - First integer
+/// * `b` - Second integer
+///
+/// # Returns
+/// The number of bytes written to the buffer or an io error
+pub fn qint_encode2<W>(cursor: &mut W, a: u32, b: u32) -> Result<usize, io::Error>
+where
+    W: Write + Seek,
+{
+    qint_encode(cursor, [a, b])
+}
+
+// Internal: Encodes one byte of using qint encoding, called in a loop.
+#[inline(always)]
+fn qint_encode_stepwise<W>(
+    leading: &mut u8,
+    cursor: &mut W,
+    mut value: u32,
+    offset: u32,
+) -> Result<usize, io::Error>
+where
+    W: Write + Seek,
+{
+    let mut ret: usize = 0;
+    let mut num_bytes: i8 = -1;
+    loop {
+        cursor.write_all(&[value as u8])?;
+        ret += &[value as u8].len();
```

Comment:
This is always going to be `ret += 1` given the way you wrote it (i.e. `&[value as u8]` is always a slice with one element).
Is that what you intended?